Source/WebCore/platform/ScrollAnimationKeyboard.cpp

```diff
@@ -0,0 +1,249 @@
+    
+    m_scrollTriggeringKeyIsPressed = false;
+
+    if (immediate) {
+        didEnd();
+        m_velocity = { };
+    }
+}
+
+bool ScrollAnimationKeyboard::animateScroll(MonotonicTime currentTime)
+{
+    auto force = FloatSize { };
+    auto axesToApplySpring = FloatSize { 1, 1 };
+    KeyboardScrollParameters params = KeyboardScrollParameters::parameters();
+
+    if (m_currentKeyboardScroll) {
+        auto scrollableDirections = scrollableDirectionsFromPosition(m_currentOffset);
+        auto direction = m_currentKeyboardScroll->direction;
+
+        if (scrollableDirections.at(boxSideForDirection(direction))) {
+            // Apply the scrolling force. Only apply the spring in the perpendicular axis,
+            // otherwise it drags against the direction of motion.
+            axesToApplySpring = perpendicularAbsoluteUnitVector(direction);
+            force = m_currentKeyboardScroll->force;
+        } else {
+            // The scroll view cannot scroll in this direction, and is rubber-banding.
+            // Apply a constant and significant force; otherwise, the force for a
+            // single-line increment is not strong enough to rubber-band perceptibly.
+            force = unitVectorForScrollDirection(direction).scaled(params.rubberBandForce);
+        }
+
+        if (fabs(m_velocity.width()) >= fabs(m_currentKeyboardScroll->maximumVelocity.width()))
+            force.setWidth(0);
+
+        if (fabs(m_velocity.height()) >= fabs(m_currentKeyboardScroll->maximumVelocity.height()))
+            force.setHeight(0);
+    }
+
+    auto extents = m_client.scrollExtentsForAnimation(*this);
+
+    FloatPoint idealPosition = (IntPoint(m_currentKeyboardScroll ? m_currentOffset : m_idealPosition).constrainedBetween(IntPoint(extents.minimumScrollOffset()), IntPoint(extents.maximumScrollOffset())));
+    FloatSize displacement = m_currentOffset - idealPosition;
+
+    auto springForce = -displacement.scaled(params.springStiffness) - m_velocity.scaled(params.springDamping);
+    force += springForce * axesToApplySpring;
+
+    float frameDuration = (currentTime - m_timeAtLastFrame).value();
+    m_timeAtLastFrame = currentTime;
+
+    FloatSize acceleration = force.scaled(1. / params.springMass);
+    m_velocity += acceleration.scaled(frameDuration);
+    m_currentOffset = m_currentOffset + m_velocity.scaled(frameDuration);
+
+    m_client.scrollAnimationDidUpdate(*this, m_currentOffset);
+
+    // Stop the spring if it reaches the ideal position.
+    FloatSize newDisplacement = m_currentOffset - idealPosition;
+
+    if (axesToApplySpring.width() && displacement.width() * newDisplacement.width() < 0)
+        m_velocity.setWidth(0);
+    if (axesToApplySpring.height() && displacement.height() * newDisplacement.height() < 0)
+        m_velocity.setHeight(0);
+
+    if (!m_scrollTriggeringKeyIsPressed && m_velocity.diagonalLengthSquared() < 1) {
+        didEnd();
+        m_velocity = { };
+    }
+
+    return true;
```

Comment:
Yes, this is because `animateScroll` overrides a virtual function, so it is forced to return a value, and in the case of `ScrollAnimationKeyboard`, the value doesn't need to be used